Add support for .to() for NestedTensor backends

ani300 opened the pull request:
Summary: This commit adds support for moving NestedTensors from CPU to GPU and back. The implementation includes requires implementing empty_like(), which is based on PR#83140.

Test Plan: Added a new unit test based on the unit test for the main .to() implementation. All unit tests must pass, as well as every sandcastle job.

Differential Revision: D40437585

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -1376,5 +1380,109 @@ Tensor reshape_as_nested(const Tensor& self, const Tensor& other) {
   return self.reshape(sizes);
 }
 
+TensorOptions verify_empty_parameters(
```
can you move this work to NestedTensorUtils.cpp

If you wanted to be a really good steward you could create a new file could NestedTensorFactories.cpp and NestedTensorFactories.h but not needed for this PR

drisspg commented on aten/src/ATen/native/nested/NestedTensorFactories.cpp:
```diff
@@ -0,0 +1,113 @@
+
+  Tensor r;
+  r = at::empty_like(self, dtype, layout, device, pin_out, memory_format);
+  if (!self.is_contiguous()) {
```
I don't know if this should mutate self. Self needs to be contiguous to call empty_like so I think so maybe just add a TORCH_CHECK

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -1,5 +1,9 @@
 #include <ATen/native/nested/NestedTensorMath.h>
 
+#include <c10/core/CPUAllocator.h>
```
do we still need these includes?

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.h:
```diff
@@ -1,7 +1,9 @@
 #pragma once
 
 #include <ATen/core/ATen_fwd.h>
+#include <ATen/ATen.h>
```
same here

drisspg commented on test/test_nestedtensor.py:
```diff
@@ -300,6 +300,72 @@ def test_nested_namespace(self):
         self.assertEqual(result, nested_namespace_result)
 
+    # Helper function to generate a random nested tensor
+    def random_nt(self, device, dtype, num_tensors, max_dims, min_dims=None):
```
Isn't there already a helper defined like this in this test file. Maybe it is just a class method and if so feel free to extract out a layer so that you can access it

drisspg commented on test/test_nestedtensor.py:
```diff
@@ -300,6 +300,72 @@ def test_nested_namespace(self):
+            ts1.append(t1)
+        return torch.nested.nested_tensor(ts1, device=device, dtype=dtype)
+
+    @torch.inference_mode()
```
Does this only work in inference_mode?

drisspg commented on test/test_nestedtensor.py:
```diff
@@ -1544,6 +1610,39 @@ def rand_mask(size):
+    @dtypes(torch.float, torch.float16, torch.double)
+    def test_empty_like(self, device, dtype):
+        ntensors = 4
+        nt = self.random_nt(device, dtype, ntensors, (4, 4))
```
I would use get contiguous and non_contiguous and make sure that non_contiguous errors

drisspg reviewed (commented):
Few small comments but otherwise I think this is very close, great job

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -1,5 +1,7 @@
 #include <ATen/native/nested/NestedTensorMath.h>
 
+#include <c10/util/Optional.h>
```
Nit: do we need to add these headers? maybe we should have had the optional.h , but do we need the Copy.h header

drisspg reviewed (approved):
👍

Commit:
Add support for .to() for NestedTensor backends (#87146)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/87146

This commit adds support for moving NestedTensors from CPU to GPU and back. The implementation includes requires implementing empty_like(), which is based on PR#83140.

Test Plan: Added a new unit test based on the unit test for the main .to() implementation. All unit tests must pass, as well as every sandcastle job.

Reviewed By: drisspg

Differential Revision: D40437585

fbshipit-source-id: b2b70ba4852c3b4ba4ae5d9ce796a37a902fb80a

ani300 commented:
@pytorchbot merge
